WIP: Update hardhat-ganache to new ganache npm module

joelamouche opened the pull request:
<!--
Thank you for using Hardhat and taking the time to send a Pull Request!

If you are introducing a new feature, please discuss it in an Issue or with someone from the team before submitting your change.

Please:
 - consider the checklist items below
 - keep the ones that make sense for your PR, and
 - DELETE the items that DON'T make sense for your PR.
-->

- [x] Because this PR includes a **bug fix**, relevant tests have been included.
- [ ] Because this PR includes a **new feature**, the change was previously discussed on an Issue or with someone from the team.
- [ ] I didn't do anything of this.

---

<!-- Add a description of your PR here -->

Handles https://github.com/NomicFoundation/hardhat/issues/3556

- Replaces the old ganache-core module with ganache.
- Changes default PORT to 8545
- Create function input is typed HardhatGanacheOptions instead of GanacheOptions in order not to confuse with the require input for the ganache server (`ServerOptions<"ethereum">`)
- _validateAndTransformOptions takes HardhatGanacheOptions input, verifies them and return correct input for the new ganache server creation (`ServerOptions<"ethereum">`) (this is not complete because it lacks a few custom parameters that are hardcoded to default, but it should work with the tests). Also remove the snakecase part from this function because this is not necessary with the new module
- Remove _registerSystemErrorsHandlers because the new package doesn't seem to support it
- Uses a simplified getDefaultOptions because the ganache package already has many default options

Some tests aren't passing yet.
I also need to write the function that supports proper custom network option conversion.

joelamouche commented on packages/hardhat-ganache/test/index.ts:
```diff
@@ -51,16 +51,15 @@ describe.skip("Ganache plugin with empty configs", function () {
-// Note: Skipping as this plugin hasn't been updated to the latest Ganache version,
-//   so it doesn't work well with Node 18
-describe.skip("Ganache plugin with custom configs", function () {
+// Note: Each of the skipped test is working when ran with .only
```
@fvictorio Please help me out on this. I don't understand why these tests work when ran alone but not with the other tests.

The first two skipped tests say `HardhatError: HH5: HardhatContext is not created.` and the last one `AssertionError: expected 1 to equal +0`

joelamouche replied:
@fvictorio they do work when ran alone

joelamouche commented on packages/hardhat-ganache/src/ganache-service.ts:
```diff
@@ -236,40 +285,86 @@ export class GanacheService {
-      }
-    });
+    // TODO: make a function that correctly does this conversion:
+    return {
```
@fvictorio Adding all the custom options is the main task that is left to do but the rest works. Please take a first pass a reviewing my work when you get the chance

fvictorio replied:
Hey @joelamouche, I'll be a bit busy the next couple of days, but I'll try to take a look next week.

fvictorio commented:
@joelamouche can you give me more context about what you've done so far in the PR? Some things that would be useful to know:

- How the API of ganache change between the old `ganache-core` module we were using and the new `ganache` module
- How this affects our implementation
- What has been your strategy so far to make this change

In general, anything that makes it easier to review the PR (like splitting into multiple commits) would be appreciated.

joelamouche commented:
@fvictorio I added in the PR description all the changes I made in this PR. I'm just adapting the current implementation to the new API which is very similar just a few differences in the way the server is started and the input typing. 

Please run the test and let me know if you get the same errors I get in the comments I added to the PR. I'm not sure they are related to the changes I made since it says `HardhatError: HH5: HardhatContext is not created.`

joelamouche commented:
Also @fvictorio when I run yarn lint I get this error:

``
``
=============

WARNING: You are currently running a version of TypeScript which is not officially supported by @typescript-eslint/typescript-estree.

You may find that it works just fine, or you may not.

SUPPORTED TYPESCRIPT VERSIONS: >=3.3.1 <4.4.0

YOUR TYPESCRIPT VERSION: 4.5.5

Please only submit bug reports when using the officially supported version.

=============

/Users/antoineestienne/Github_Files/hardhat/packages/hardhat-ganache/src/ganache-service.ts
    4:1   error  'hardhat/src/internal/core/errors' import is restricted from being used by a pattern                  no-restricted-imports
  155:20  error  Unexpected nullable number value in conditional. Please handle the nullish/zero/NaN cases explicitly  @typescript-eslint/strict-boolean-expressions
  156:24  error  Unexpected nullable string value in conditional. Please handle the nullish/empty cases explicitly     @typescript-eslint/strict-boolean-expressions
  179:19  error  Expected '!==' and instead saw '!='                                                                   eqeqeq
  216:47  error  'reject' is defined but never used. Allowed unused args must match /^_/u                              @typescript-eslint/no-unused-vars

/Users/antoineestienne/Github_Files/hardhat/packages/hardhat-ganache/src/index.ts
  2:1  error  'hardhat/src/builtin-tasks/task-names' import is restricted from being used by a pattern  no-restricted-imports
  3:1  error  'hardhat/src/config' import is restricted from being used by a pattern                    no-restricted-imports
  4:1  error  'hardhat/src/types' import is restricted from being used by a pattern                     no-restricted-imports

/Users/antoineestienne/Github_Files/hardhat/packages/hardhat-ganache/test/helpers.ts
  1:1  error  'hardhat/src/plugins-testing' import is restricted from being used by a pattern  no-restricted-imports
  2:1  error  'hardhat/src/types' import is restricted from being used by a pattern            no-restricted-imports

✖ 10 problems (10 errors, 0 warnings)
````

But the fixes are not made.
As for the typescript error, it is using the version that is in the package.json Should we lock typescript and @typescript-eslint/eslint-plugin versions?

alcuadrado commented:
Thanks for submitting this PR! Unfortunately we've decided to removed some packages from the monorepo, and `hardhat-ganache` is one of them.

joelamouche commented:
@alcuadrado So has it been moved to another repo? Is the feature going to be implemented?

alcuadrado commented:
It hasn't been moved to a separate repository. We can do it and reconsider the upgrade if there's enough interest. 

That being said, we hadn't released a version of it for almost two years, and got really little interest on it.
